features/flows/classification/engine/impl/src/main/java/org/opennms/netmgt/flows/classification/internal/DefaultClassificationEngine.java

```diff
@@ -50,20 +53,65 @@
     private final Comparator<RuleDefinition> ruleComparator = new RulePositionComparator();
     private final ClassificationRuleProvider ruleProvider;
-    private final FilterService filterService;
 
-    public DefaultClassificationEngine(ClassificationRuleProvider ruleProvider, FilterService filterService) {
+    public DefaultClassificationEngine(final ClassificationRuleProvider ruleProvider, final FilterService filterService) {
         this(ruleProvider, filterService, true);
     }
 
-    public DefaultClassificationEngine(ClassificationRuleProvider ruleProvider, FilterService filterService, boolean initialize) {
+    public DefaultClassificationEngine(final ClassificationRuleProvider ruleProvider, final FilterService filterService, final boolean initialize) {
         this.ruleProvider = Objects.requireNonNull(ruleProvider);
-        this.filterService = Objects.requireNonNull(filterService);
+        this.portClassifiersCache = CacheBuilder.newBuilder().build(new CacheLoader<Integer, List<Classifier>>() {
+            @Override
+            public List<Classifier> load(Integer port) throws Exception {
+                final List<RuleDefinition> ruleDefinitions = rulePortList.get(port);
+                final List<Classifier> classifiers = ruleDefinitions.stream().map(rule -> {
+                    if (!ruleClassifierMap.containsKey(rule)) {
+                        final DefaultRuleDefinition portRule = new DefaultRuleDefinition();
+                        portRule.setName(rule.getName());
+                        portRule.setProtocol(rule.getProtocol());
+                        portRule.setSrcAddress(rule.getSrcAddress());
+                        portRule.setDstAddress(rule.getDstAddress());
+                        portRule.setExporterFilter(rule.getExporterFilter());
+                        portRule.setGroupPosition(rule.getGroupPosition());
+                        portRule.setPosition(rule.getPosition());
+
+                        // Check weather to apply rule for src or dst port (both may be very unlikely, but possible)
+                        if (rule.hasDstPortDefinition() && rule.hasSrcPortDefinition()) {
+                            portRule.setSrcPort(rule.getSrcPort()); // keep src port as is, to apply filter
+                        } else {
+                            // Only src or dst ports are defined (or none)
+                            // if none, the value of either src or dst port may be empty, as the filtering already occurred
+                            // through the index of the rule in the classifierPortList.
+                            if (rule.hasDstPortDefinition()) {
+                                portRule.setDstPort(rule.getDstPort());
+                            }
+                            if (rule.hasSrcPortDefinition()) {
+                                portRule.setSrcPort(rule.getSrcPort());
+                            }
+                        }
```

Comment:
Am I missing something? It looks like if the rule has a dst and src, it only set a SrcPort and ignores DstPort. Shouldn't just the code in the else be present as it would work on all cases?

Reply:
Yes, you are absolutely right. Before the change it was necessary to distinguish between both cases, now it can be merged.